polkadot/xcm/xcm-executor/src/traits/transact_asset.rs

```diff
@@ -249,10 +269,18 @@ impl TransactAsset for Tuple {
 		)* );
 	}
 
-	fn deposit_asset(what: &Asset, who: &Location, context: Option<&XcmContext>) -> XcmResult {
+	fn deposit_asset(
+		mut what: AssetsInHolding,
+		who: &Location,
+		context: Option<&XcmContext>,
+	) -> Result<(), (AssetsInHolding, XcmError)> {
 		for_tuples!( #(
 			match Tuple::deposit_asset(what, who, context) {
-				Err(XcmError::AssetNotFound) | Err(XcmError::Unimplemented) => (),
+				// Err((unspent, error)) if error == XcmError::AssetNotFound || error == XcmError::Unimplemented => (),
+				Err((unspent, XcmError::AssetNotFound)) | Err((unspent, XcmError::Unimplemented)) => {
+					what = unspent;
```

Comment:
should `what` be just the last? or sum?

same comment applies to `deposit_asset_with_surplus`

Reply:
It should be the last since the transactors don't resolve/drop it when throwing an error. The same imbalance has to be passed from the first element of the tuple all the way to the end as long as they keep failing.